hazelcast/src/main/java/com/hazelcast/map/impl/nearcache/NearCacheInvalidatorImpl.java

```diff
@@ -0,0 +1,413 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.hazelcast.instance.GroupProperty.MAP_INVALIDATION_MESSAGE_BATCH_ENABLED;
+import static com.hazelcast.instance.GroupProperty.MAP_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS;
+import static com.hazelcast.instance.GroupProperty.MAP_INVALIDATION_MESSAGE_BATCH_SIZE;
+import static com.hazelcast.map.impl.MapService.SERVICE_NAME;
+import static com.hazelcast.util.CollectionUtil.isEmpty;
+import static com.hazelcast.util.ConcurrencyUtil.getOrPutIfAbsent;
+
+/**
+ * Default implementation of {@link NearCacheInvalidator}
+ *
+ * @see NearCacheInvalidator
+ */
+public class NearCacheInvalidatorImpl implements NearCacheInvalidator {
+
+    private final ConstructorFunction<String, InvalidationQueue> invalidationQueueConstructor
+            = new ConstructorFunction<String, InvalidationQueue>() {
+        @Override
+        public InvalidationQueue createNew(String mapName) {
+            return new InvalidationQueue();
+        }
+    };
+
+    private final ConstructorFunction<String, Set<String>> registryConstructor
+            = new ConstructorFunction<String, Set<String>>() {
+        @Override
+        public Set<String> createNew(String mapName) {
+            return Collections.newSetFromMap(new ConcurrentHashMap());
+        }
+    };
+
+    private final boolean batchingEnabled;
+    private final int batchSize;
+    private final ConcurrentMap<String, InvalidationQueue> invalidationQueueMap
+            = new ConcurrentHashMap<String, InvalidationQueue>();
+    private final ConcurrentMap<String, Set<String>> clientInvalidationListenerIds
+            = new ConcurrentHashMap<String, Set<String>>();
+    private final EventService eventService;
+    private final NodeEngine nodeEngine;
+    private final MapServiceContext mapServiceContext;
+    private final NearCacheProvider nearCacheProvider;
+
+    NearCacheInvalidatorImpl(MapServiceContext mapServiceContext, NearCacheProvider nearCacheProvider) {
+        this.mapServiceContext = mapServiceContext;
+        this.nearCacheProvider = nearCacheProvider;
+        this.nodeEngine = mapServiceContext.getNodeEngine();
+        this.eventService = nodeEngine.getEventService();
+        GroupProperties groupProperties = nodeEngine.getGroupProperties();
+        this.batchSize = groupProperties.getInteger(MAP_INVALIDATION_MESSAGE_BATCH_SIZE);
+        this.batchingEnabled = groupProperties.getBoolean(MAP_INVALIDATION_MESSAGE_BATCH_ENABLED) && batchSize > 1;
+        int invalidationMessageBatchFreq = groupProperties.getInteger(MAP_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS);
+        nodeEngine.getExecutionService()
```

Comment:
Even though batching is not enabled, you schedule a task. There should be a check